gms/gossiper.cc

```diff
@@ -557,6 +583,85 @@ void gossiper::notify_failure_detector(inet_address endpoint, const endpoint_sta
+                endpoint_state_map[node] = remote_state;
+            }
+        } else if (remote_generation == local_generation) {
+            if (listener_notification) {
+                // find maximum state
+                int local_max_version = this->get_max_endpoint_state_version(local_state);
+                int remote_max_version = this->get_max_endpoint_state_version(remote_state);
+                if (remote_max_version > local_max_version) {
+                    // apply states, but do not notify since there is no major change
+                    this->apply_new_states(node, local_state, remote_state);
+                } else {
+                    logger.trace("Ignoring remote version {} <= {} for {}", remote_max_version, local_max_version, node);
+                }
+                if (!local_state.is_alive() && !this->is_dead_state(local_state)) { // unless of course, it was dead
+                    this->mark_alive(node, local_state);
+                }
+            } else {
+                for (const auto& item : remote_state.get_application_state_map()) {
+                    const auto& remote_key = item.first;
+                    const auto& remote_value = item.second;
+                    const versioned_value* local_value = local_state.get_application_state_ptr(remote_key);
+                    if (!local_value || remote_value.version > local_value->version) {
+                        logger.debug("Applying remote_state for node {} (remote generation = local generation), key={}, value={}",
+                                node, remote_key, remote_value);
+                        local_state.add_application_state(remote_key, remote_value);
+                    } else {
+                        logger.debug("Ignoring remote_state for node {} (remote generation = local generation), key={}, value={}", node, remote_key, remote_value);
+                    }
+                }
+            }
+        } else {
+            logger.debug("Ignoring remote generation {} < {}", remote_generation, local_generation);
+        }
+    } else {
+        if (listener_notification) {
+            // this is a new node, report it to the FD in case it is the first time we are seeing it AND it's not alive
+            fd().report(node);
+            this->handle_major_state_change(node, remote_state);
+        } else {
+            logger.debug("Applying remote_state for node {} (new node)", node);
+            endpoint_state_map[node] = remote_state;
+        }
+    }
+}
+
+// Runs inside seastar::async context
+void gossiper::apply_state_locally_without_listener_notification(std::unordered_map<inet_address, endpoint_state> map) {
+    for (auto& x : map) {
+        const inet_address& node = x.first;
+        const endpoint_state& remote_state = x.second;
+        if (node == this->get_broadcast_address()) {
+            continue;
+        }
+        do_apply_state_locally(node, remote_state, false);
```

Comment:
When do we notify the listeners? Can't we create a situation where we learn something from the cluster, but lost the infomation because we don't notify?

Reply:
This is for shadow round only. After gossip service is started normally, we will learn everything and do the notification properly.